### PR TITLE
USF-1327: Checkout containers (redux)

### PR DESCRIPTION
**Pull request by jeff-matthews**
This pull request (PR) documents Checkout container configuration options based on the [doc template](https://github.com/commerce-docs/microsite-commerce-storefront/blob/develop/_dropin-templates/dropin-containers.mdx?plain=1) for dropin containers.

Options and descriptions are based on the prop interfaces defined for each component in the source code.

### CODE REVIEW

File: src/content/docs/dropins/checkout/containers/estimate-shipping.mdx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  compact
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['estimated', 'boolean', 'No', 'A boolean indicating whether the shipping cost is estimated.'],
+    ['price', 'VNode', 'Yes', 'The rendered price component, which can be free shipping, the price including tax, or the price excluding tax.'],
+    ['priceExclTax', 'VNode', 'No', 'The rendered price component excluding tax, if applicable.'],
+    ['taxExcluded', 'boolean', 'No', 'A boolean indicating whether the tax is excluded from the shipping price.'],
+    ['taxIncluded', 'boolean', 'No', 'A boolean indicating whether the tax is included in the shipping price.'],
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/estimate-shipping.mdx]: Although `EstimateShipping` is a checkout container, it's a bit special. We had to create this one to overwrite the `EstimateShipping` defined into the Storefront Cart dropin. Whereas the `EstimateShipping` defined into the Storefront Cart dropin is editable, the `EstimateShipping` defined into the Checkout dropin is readonly.
Initially it shows the estimate shipping costs, but once the customer has introduced a shipping address and a shipping method is selected, then it shows the real shipping cost.
You can see the estimate shipping information inside of the `OrderSummary` container (comming from the Storefront Cart dropin)... so it only makes sense to use it as slot of the `OrderSummary` container.

**example**:

```
import { OrderSummary } from '@dropins/storefront-cart/containers/OrderSummary.js';
import { render as CartProvider } from '@dropins/storefront-cart/render.js';

const $orderSummary = checkoutFragment.querySelector(
  '.checkout__order-summary'
);

CartProvider.render(OrderSummary, {
  slots: {
    EstimateShipping: (esCtx) => {
      const estimateShippingForm = document.createElement('div');
      CheckoutProvider.render(EstimateShipping)(estimateShippingForm);
      esCtx.appendChild(estimateShippingForm);
    },
  },
})($orderSummary),
```

File: src/content/docs/dropins/checkout/containers/login-form.mdx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    ['hint', 'string', 'No', 'A hint or placeholder text for the email input field.'],
+    ['error', 'string | null', 'No', 'Any error message related to the email input.'],
+    ['onEmailChange', 'function', 'Yes', 'A function that handles changes to the email input field. It takes an event as an argument.'],
+    ['onEmailBlur', 'function', 'Yes', 'A function that handles the blur event (when the email input loses focus). It takes an event as an argument.'],
+    ['onEmailInvalid', 'function', 'Yes', 'A function that handles invalid email input. It takes an event as an argument.'],
+    ['onSignInClick', 'function', 'No', 'A function that handles the sign-in button click. It takes the email (string or null) as an argument.'],
+    ['onSignOutClick', 'function', 'No', 'A function that handles the sign-out button click. It takes no arguments.'],
+    ['customer', 'Customer | null', 'No', 'The customer information, if available.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/login-form.mdx]: `LoginForm` uses the `AuthCombine` container defined into the Storefront Auth dropin.

**example**:

```
import AuthCombine from '@dropins/storefront-auth/containers/AuthCombine.js';
import { render as AuthProvider } from '@dropins/storefront-auth/render.js';

import LoginForm from '@dropins/storefront-checkout/containers/LoginForm.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $login = checkoutFragment.querySelector('.checkout__login');

CheckoutProvider.render(LoginForm, {
  name: LOGIN_FORM_NAME,
  onSignInClick: async (initialEmailValue) => {
    const signInForm = document.createElement('div');

    AuthProvider.render(AuthCombine, {
      signInFormConfig: {
        renderSignUpLink: true,
        initialEmailValue,
        onSuccessCallback: () => {
          displayOverlaySpinner();
        },
      },
      signUpFormConfig: {},
      resetPasswordFormConfig: {},
    })(signInForm);

    showModal(signInForm);
  },
  onSignOutClick: () => {
    authApi.revokeCustomerToken();
  },
})($login),
```

File: src/content/docs/dropins/checkout/containers/out-of-stock.mdx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+<OptionsTable
+  compact
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['items', 'CartItem[]', 'Yes', 'The list of items that are out of stock.'],
+    ['onRemoveOutOfStock', 'function', 'No', 'Handles the removal of out-of-stock items. It takes an event as an argument.'],
+    ['routeCart', 'string', 'No', 'The route to the cart page.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/out-of-stock.mdx]: **example**:

```
import OutOfStock from '@dropins/storefront-checkout/containers/OutOfStock.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $outOfStock = checkoutFragment.querySelector('.checkout__out-of-stock');

CheckoutProvider.render(OutOfStock, {
  routeCart: () => '/cart',
  onCartProductsUpdate: (items) => {
    cartApi.updateProductsFromCart(items).catch(console.error);
  },
})($outOfStock),
```

File: src/content/docs/dropins/checkout/containers/payment-methods.mdx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+The `PaymentMethods` container provides the following configuration options:
+
+<OptionsTable
+  compact
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['slots.Main', 'SlotProps<PaymentMethodsMainSlotContext>', 'No', 'The main slot for the payment methods.'],
+    ['slots.Handlers', 'PaymentMethodHandlerSlots', 'No', 'The handlers for different payment methods.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/payment-methods.mdx]: **example**:

```
import PaymentMethods from '@dropins/storefront-checkout/containers/PaymentMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $paymentMethods = checkoutFragment.querySelector(
  '.checkout__payment-methods'
);

CheckoutProvider.render(PaymentMethods)($paymentMethods),
```

File: src/content/docs/dropins/checkout/containers/place-order.mdx [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+The `PlaceOrder` container provides the following configuration options:
+
+<OptionsTable
+  compact
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['onPlaceOrder', 'function', 'No', 'Handles the place order action. It takes a PlaceOrderContext object as an argument and returns a Promise<void>.'],
+    ['disabled', 'boolean', 'No', 'Indicates whether the place order button is disabled.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/place-order.mdx]: `PlaceOrder` container accepts a function as a prop to be used to perform some validations (like forms, etc):

```handleValidation?: () => boolean;```

**example**:

```
import PlaceOrder from '@dropins/storefront-checkout/containers/PlaceOrder.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $placeOrder = checkoutFragment.querySelector('.checkout__place-order');

CheckoutProvider.render(PlaceOrder, {
  handleValidation: () => {
    let success = true;
    const forms = document.forms;

    const loginForm = forms[LOGIN_FORM_NAME];

    if (loginForm) {
      success = loginForm.checkValidity();
      if (!success) scrollToElement($login);
    }

    const shippingForm = forms[SHIPPING_FORM_NAME];

    if (
      success &&
      shippingFormRef.current &&
      shippingForm &&
      shippingForm.checkVisibility()
    ) {
      success = shippingFormRef.current.handleValidationSubmit(false);
    }

    const billingForm = forms[BILLING_FORM_NAME];

    if (
      success &&
      billingFormRef.current &&
      billingForm &&
      billingForm.checkVisibility()
    ) {
      success = billingFormRef.current.handleValidationSubmit(false);
    }

    return success;
  },
  onPlaceOrder: async () => {
    displayOverlaySpinner();

    try {
      await checkoutApi.placeOrder();
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      removeOverlaySpinner();
    }
  },
})($placeOrder),
]);
```

File: src/content/docs/dropins/checkout/containers/server-error.mdx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+<OptionsTable
+  compact
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['errorMessage', 'string', 'Yes', 'The error message to display.'],
+    ['contactSupport', 'string', 'No', 'Provides a contact support message or link.'],
+    ['onClick', 'function', 'No', 'Handles click events. It takes an event (e: any) as an argument.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/server-error.mdx]: **example**:

```
import ServerError from '@dropins/storefront-checkout/containers/ServerError.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $serverError = checkoutFragment.querySelector(
  '.checkout__server-error'
);

CheckoutProvider.render(ServerError, {
  onRetry: () => {
    $content.classList.remove('checkout__content--error');
  },
  onServerError: () => {
    $content.classList.add('checkout__content--error');
  },
})($serverError),
```

File: src/content/docs/dropins/checkout/containers/shipping-methods.mdx [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  options={[
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['isLoading', 'boolean', 'No', 'Indicates whether the shipping methods are loading. Default is false.'],
+    ['onSelectionChange', 'function', 'No', 'Handles the selection of a shipping method. Takes a ShippingMethod object as an argument.'],
+    ['options', 'ShippingMethod[]', 'No', 'The available shipping methods.'],
+    ['selection', 'ShippingMethod', 'No', 'The currently selected shipping method.'],
+    ['className', 'string', 'No', 'The CSS class name(s) to apply to the component.'],
+    ['data-testid', 'string', 'No', 'Used for testing purposes to identify the component.']
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/shipping-methods.mdx]: **example**:

```
import ShippingMethods from '@dropins/storefront-checkout/containers/ShippingMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $delivery = checkoutFragment.querySelector('.checkout__delivery');

CheckoutProvider.render(ShippingMethods, {
  hideOnVirtualCart: true,
  onCheckoutDataUpdate: () => {
    cartApi.refreshCart().catch(console.error);
  },
})($delivery),
```

**Review by mcocapiza — CHANGES_REQUESTED** [review]
Also, you need to add a couple of more containers:
- BillToShippingAddress => Checkbox to decide if the billing address will be the same as the shipping address, otherwise the billing address form is shown

**example**:

```
import BillToShippingAddress from '@dropins/storefront-checkout/containers/BillToShippingAddress.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $billToShipping = checkoutFragment.querySelector(
  '.checkout__bill-to-shipping'
);

CheckoutProvider.render(BillToShippingAddress, {
  hideOnVirtualCart: true,
  onChange: (checked) => {
    $billingForm.style.display = checked ? 'none' : 'block';

    if (!checked && billingFormRef.current) {
      const isDataValid = billingFormRef.current.handleValidationSubmit();

      setAddressOnCart(
        { data: billingFormRef.current.formData, isDataValid },
        checkoutApi.setBillingAddress
      );
    }
  },
})($billToShipping),
```

- OrderConfirmationHeader => Once the order is placed, a new order confirmation page is shown with the order details. This container also has the "Create an account" button (Sign-up) in case the guest user wanted to create an account

**example**:

```
import OrderConfirmationHeader from '@dropins/storefront-checkout/containers/OrderConfirmationHeader.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $orderConfirmationHeader = orderConfirmationFragment.querySelector(
  '.order-confirmation__header'
);

CheckoutProvider.render(OrderConfirmationHeader, {
  orderData,
  onSignUpClick,
})($orderConfirmationHeader);
```

File: src/content/docs/dropins/checkout/containers/error-banner.mdx [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    ['Option', 'Type', 'Req?', 'Description'],
+    ['className', 'string', 'No', 'CSS class to apply to the container.'],
+    ['data-testid', 'string', 'No', 'Identifier for the banner for testing purposes.'],
+    ['icon', 'JSX.Element', 'No', 'Icon to display in the banner.'],
+    ['message', 'string', 'Yes', 'Message to display in the banner.'],
+    ['aria-label', 'string', 'No', 'Accessibility label for the banner.'],
+    ['onDismiss', 'function', 'No', 'Function to call when the banner is dismissed.'],
+    ['variant', 'string', 'No', 'Variant of the banner indicating the type of alert.'],
+  ]}
+/>
```
Comment (mcocapiza) [inline src/content/docs/dropins/checkout/containers/error-banner.mdx]: Please remove this block... we are not going to use it.

**Comment by jeff-matthews** [comment]
Thanks @mcocapiza. I addressed your feedback and added the additional containers (`BillToShippingAddress` and `OrderConfirmationHeader`).

Will you please provide a code sample for the `MergedCartBanner` container?

**Comment by jeff-matthews** [comment]
I found a simple example in Slack. I'll push changes in a moment.

**Comment by mcocapiza** [comment]
- `MergedCartBanner` => When the customer performs a "Sign-In" action, if he/she/it had a previous cart within products inserted, then a banner is shown informing the current items added into the guest cart and the items recovered from a previous customer cart will be merged. 

**example**:

```
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $mergedCartBanner = checkoutFragment.querySelector(
  '.checkout__merged-cart-banner',
);

CheckoutProvider.render(MergedCartBanner)($mergedCartBanner),
```

**Comment by mcocapiza** [comment]
Please update the `EstimateShipping` example (I missed to import the `EstimateShipping` container from Checkout dropin):

```
import { OrderSummary } from '@dropins/storefront-cart/containers/OrderSummary.js';
import { render as CartProvider } from '@dropins/storefront-cart/render.js';
import EstimateShipping from '@dropins/storefront-checkout/containers/EstimateShipping.js';

const $orderSummary = checkoutFragment.querySelector(
  '.checkout__order-summary'
);

CartProvider.render(OrderSummary, {
  slots: {
    EstimateShipping: (esCtx) => {
      const estimateShippingForm = document.createElement('div');
      CheckoutProvider.render(EstimateShipping)(estimateShippingForm);
      esCtx.appendChild(estimateShippingForm);
    },
  },
})($orderSummary),

```

**Comment by mcocapiza** [comment]
Please update the `OutOfStock` example (I missed to import the `cartApi` from Cart dropin):

```
import * as cartApi from '@dropins/storefront-cart/api.js';
import OutOfStock from '@dropins/storefront-checkout/containers/OutOfStock.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $outOfStock = checkoutFragment.querySelector('.checkout__out-of-stock');

CheckoutProvider.render(OutOfStock, {
  routeCart: () => '/cart',
  onCartProductsUpdate: (items) => {
    cartApi.updateProductsFromCart(items).catch(console.error);
  },
})($outOfStock),
```

**Comment by mcocapiza** [comment]
Please update the `PlaceOrder` example (I missed to import the `checkoutApi` from Checkout dropin):

```
import * as checkoutApi from '@dropins/storefront-checkout/api.js';
import PlaceOrder from '@dropins/storefront-checkout/containers/PlaceOrder.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $placeOrder = checkoutFragment.querySelector('.checkout__place-order');

CheckoutProvider.render(PlaceOrder, {
  handleValidation: () => {
    let success = true;
    const forms = document.forms;

    const loginForm = forms[LOGIN_FORM_NAME];

    if (loginForm) {
      success = loginForm.checkValidity();
      if (!success) scrollToElement($login);
    }

    const shippingForm = forms[SHIPPING_FORM_NAME];

    if (
      success &&
      shippingFormRef.current &&
      shippingForm &&
      shippingForm.checkVisibility()
    ) {
      success = shippingFormRef.current.handleValidationSubmit(false);
    }

    const billingForm = forms[BILLING_FORM_NAME];

    if (
      success &&
      billingFormRef.current &&
      billingForm &&
      billingForm.checkVisibility()
    ) {
      success = billingFormRef.current.handleValidationSubmit(false);
    }

    return success;
  },
  onPlaceOrder: async () => {
    displayOverlaySpinner();

    try {
      await checkoutApi.placeOrder();
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      removeOverlaySpinner();
    }
  },
})($placeOrder),
]);
```

**Comment by mcocapiza** [comment]
The `ServerError` configurations are the following ones:
- onRetry?: () => void; /* what to do in case of retry */
- onServerError?: (error: string) => void; /* what to do in case of a server error */

**Comment by mcocapiza** [comment]
Please update the `ShippingMethods` example (I missed to import the `cartApi` from Cart dropin): 

```
import * as cartApi from '@dropins/storefront-cart/api.js';
import ShippingMethods from '@dropins/storefront-checkout/containers/ShippingMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $delivery = checkoutFragment.querySelector('.checkout__delivery');

CheckoutProvider.render(ShippingMethods, {
  hideOnVirtualCart: true,
  onCheckoutDataUpdate: () => {
    cartApi.refreshCart().catch(console.error);
  },
})($delivery),
```

**Comment by mcocapiza** [comment]
The `ShippingMethods` configurations are the following ones:
- onCheckoutDataUpdate?: () => Promise<void>; /* what to do in case the checkout data has been updated */
- onShippingMethodSelect?: (method: ShippingMethod) => void; /* what to do in case a shipping method has been selected */
- preSelectedMethod?: { carrierCode: string; methodCode: string }; /* if we want to preselect a specific shipping method */
- shippingMethodsSlot?: SlotProps; /* to provide more shipping methods */

Also we could include another prop:
- hideOnVirtualCart: boolean No Hides the container if the cart is virtual.

**Comment by jeff-matthews** [comment]
Thanks @mcocapiza. All revisions applied.

**Review by keharper — APPROVED** [review]

**Pull request merged**